File: src/core/src/main/kotlin/com/cshep4/premierpredictor/component/team/FormFormatter.kt
package com.cshep4.premierpredictor.component.team

import com.cshep4.premierpredictor.data.Match
import com.cshep4.premierpredictor.data.TeamForm
import com.cshep4.premierpredictor.data.TeamMatchResult
import com.cshep4.premierpredictor.enum.Location
import com.cshep4.premierpredictor.enum.Location.AWAY
import com.cshep4.premierpredictor.enum.Location.HOME
import com.cshep4.premierpredictor.enum.Result
import com.cshep4.premierpredictor.enum.Result.*
import org.springframework.stereotype.Component

@Component
class FormFormatter {
    fun formatLastFiveGames(teamMatches: Map<String, List<Match>>): Map<String, TeamForm> {
        return teamMatches.map { Pair(it.key, createTeamForm(it.key, it.value.takeLast(5))) }.toMap()
    }

    private fun createTeamForm(teamName: String, matches: List<Match>): TeamForm {
        val matchResults = matches.map { convertMatchToResultSummary(teamName, it) }

        return TeamForm(form = matchResults)
    }

    private fun convertMatchToResultSummary(teamName: String, match: Match): TeamMatchResult {
        return TeamMatchResult(result = getResult(teamName, match), score = getScore(match), opponent = getOpponent(teamName, match), location = getLocation(teamName, match))
    }

    private fun getResult(teamName: String, match: Match): Result {
        if (match.hGoals == match.aGoals) {
            return DRAW
        }

        if (getLocation(teamName, match) == HOME) {
            if (match.hGoals!! > match.aGoals!!) {
                return WIN
            }
        } else {
            if (match.aGoals!! > match.hGoals!!) {
                return WIN
            }
        }

        return LOSS
    }

    private fun getScore(match: Match): String = "${match.hGoals}-${match.aGoals}"

    private fun getOpponent(teamName: String, match: Match): String {
        return when (teamName) {
            match.hTeam -> match.aTeam
            else -> match.hTeam
        }
    }

    private fun getLocation(teamName: String, match: Match): Location {
        return when (teamName) {
            match.hTeam -> HOME
            else -> AWAY
        }
    }
}

File: src/fixtureservice/lib/component/form-formatter.ts
import {Fixture} from "../model/fixture";
import {Result, TeamForm, TeamMatchResult, Location} from "../model/form";

const {HOME, AWAY} = Location;
const {WIN, DRAW, LOSS} = Result;

export class FormFormatter {
    public formatLastFiveGames(teamFixtures: Map<string, Fixture[]>): Promise<Map<string, TeamForm>> {
        return new Promise<Map<string, TeamForm>>((resolve, reject) => {
            const forms: Map<string, TeamForm> = new Map<string, TeamForm>();

            teamFixtures.forEach((v: Fixture[], k: string) => {
                forms.set(k, this.createTeamForm(k, v.slice(-5)));
            });

            resolve(forms);
        });
    }

    private createTeamForm(teamName: String, fixtures: Fixture[]): TeamForm {
        return {
            forms: fixtures.map(f => this.convertMatchToResultSummary(teamName, f))
        }
    }

    private convertMatchToResultSummary(teamName: String, fixture: Fixture): TeamMatchResult {
        return {
            result: this.getResult(teamName, fixture),
            score: this.getScore(fixture),
            opponent: this.getOpponent(teamName, fixture),
            location: this.getLocation(teamName, fixture)
        }
    }

    private getResult(teamName: String, fixture: Fixture): Result {
        if (fixture.hGoals == fixture.aGoals) {
            return DRAW
        }

        if (this.getLocation(teamName, fixture) == HOME) {
            if (fixture.hGoals > fixture.aGoals) {
                return WIN
            }
        } else {
            if (fixture.aGoals > fixture.hGoals) {
                return WIN
            }
        }

        return LOSS
    }

    private getScore = (fixture: Fixture) => `${fixture.hGoals}-${fixture.aGoals}`;

    private getOpponent(teamName: String, fixture: Fixture): string {
        if (fixture.hTeam === teamName) {
            return fixture.aTeam;
        }

        return fixture.hTeam;
    }

    private getLocation(teamName: String, fixture: Fixture): Location {
        if (fixture.hTeam === teamName) {
            return HOME;
        }

        return AWAY;
    }
}